Sources/Client/WebSocket/WebSocket.swift

```diff
@@ -191,12 +191,38 @@ extension WebSocket {
         }
     }
     
+    func subscribe(forEvents eventTypes: Set<EventType> = Set(EventType.allCases), callback: @escaping Client.OnEvent) -> Cancellable {
+        let subscription = Subscription { [weak self] uuid in
+            self?.webSocket.callbackQueue.async {
+                self?.onEventObservers[uuid] = nil
+            }
+        }
+        
+        let handler: Client.OnEvent = { event in
+            guard eventTypes.contains(event.type) else {
+                return
+            }
+            
+            callback(event)
+        }
+        
+        onEventObservers[subscription.uuid] = handler
+        
+        return subscription
+    }
+    
     private func clearStateAfterDisconnect() {
         handshakeTimer.suspend()
         lastConnectionId = nil
         cancelBackgroundWork()
         logger?.log("🔑 connectionId cleaned")
     }
+    
+    private func onEvent(_ event: Event) {
+        webSocket.callbackQueue.async { [weak self] in
+            self?.onEventObservers.values.forEach({ $0(event) })
+        }
+    }
```

Comment:
This is no sense now.

Reply:
Please check how it used. It needs to cleanup and probably remove `self.callbackQueue`

Reply:
`onEvent` function is needed since it's called from 2 different places, `WebSocket:L254` and `WebSocket:L262`, I don't want to replicate this call in 2 different places.
`webSocket.callbackQueue.async` is needed since `onEvent` is called in `callbackQueue`, which is different from `webSocket.callbackQueue`

Reply:
Please look closer, in WebSocket delegate (`webSocketQueue`) on event it async in `callbackQueue` and then you async to `webSocketQueue` back!

Reply:
I have looked closer, `onEvent` is called inside `performInCallbackQueue` which is:
```swift
private func performInCallbackQueue(execute block: @escaping () -> Void) {
    if let callbackQueue = callbackQueue {
        callbackQueue.async(execute: block)
    } else {
        block()
    }
}
```
which *is not* the same as `webSocket.callbackQueue`.
Inside `WebSocket`, we have:
1. `callbackQueue`, which `performInCallbackQueue` uses
2. `webSocket.callbackQueue` which I'm using for handling events.

Reply:
```swift
    func websocketDidReceiveMessage(socket: Starscream.WebSocketClient, text: String) {
        ... // here webSocket.callbackQueue
        
        if case let .healthCheck(connectionId, _) = event {
            ...
            performInCallbackQueue { [weak self] in
                self?.onEvent(event)
                self?.connection = .connected
            }
            
            return
        }
        
        if isConnected {
            performInCallbackQueue { [weak self] in self?.onEvent(event) }
        }
    }
```

`performInCallbackQueue` -> optionally switch queue
`onEvent` -> you switch back to `webSocket.callbackQueue`

Reply:
Yes, exactly, because we need to use `webSocket.callbackQueue`
And `callbackQueue` (which `performInCallbackQueue` uses) is not the same `webSocket.callbackQueue`

Reply:
The changes should be like that:
```swift
func websocketDidReceiveMessage(socket: Starscream.WebSocketClient, text: String) {
        ...
        
        if case let .healthCheck(connectionId, _) = event {
            ...
            onEventObservers.values.forEach({ $0(event) })
            
            return
        }
        
        if isConnected {
            onEventObservers.values.forEach({ $0(event) })
        }
    }
```

Reply:
Ok I totally missed
`// here webSocket.callbackQueue`
comment in the code snippet above, which clarifies your point. I'll make the necessary changes 👍🏼